app/models/export/activity_actuals_columns.rb

```diff
@@ -23,10 +23,10 @@ def headers
 
   def rows
     return [] if @activities.empty?
-
-    @activities.map { |activity|
-      actual_and_refund_data(activity)
-    }.to_h
+    @_rows ||=
+      @activities.map { |activity|
+        actual_and_refund_data(activity)
+      }.to_h
```

Comment:
did you consider describing (and protecting) this optimisation in a spec?

e.g. something along the lines of:

```rb
context "when  #rows is called multiple times" do
  before do
    allow(Export::FinancialQuarterActivityTotals).to receive(:new).and_return(fake_totals)

    3.times { subject.rows }
  end
  
  it "builds totals for actuals once only" do
    expect(Export::FinancialQuarterActivityTotals)
      .to have_received(:new)
      .with(hash_including(type: :actual))
      .once
  end
  
  it "builds totals for refunds once only" do
    expect(Export::FinancialQuarterActivityTotals)
      .to have_received(:new)
      .with(hash_including(type: :refund))
      .once
  end
end
```

Reply:
I did consider it, but decided it was testing ruby behaviour - happy to add it in though - especially as I have a way to do so now - thank you!

Reply:
@ed, what do you think about this:

```ruby
    context "when #rows is called multiple times" do
      let(:breakdown) { false }
      let(:report) { nil }

      before do
        totals_double = double(call: {"fake_activity_id" => []})
        allow(Export::AllActivityTotals).to receive(:new).and_return(totals_double)

        3.times { subject.rows }
      end

      it "builds totals for the activities only once" do
        expect(Export::AllActivityTotals)
          .to have_received(:new)
          .at_least(@activities.count).times
      end
    end
```

`Export::AllActivityTotals` is called on each activity, so in the spec 5 activites means 5 calls when cached?

Reply:
👍 makes sense, very clear but I think `at_least(n)` is too lax? That will be as happy with 15 times as with 5 times. Can we be precise, e.g. `.exactly(5).times` to demonstrate that the expensive operation are performed only once?

Reply:
🤦 yep, I meant exactly and was looking for the syntax, I meant to change it! 👍